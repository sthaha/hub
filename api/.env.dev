ENVIRONMENT="development"

POSTGRESQL_HOST="localhost"
POSTGRESQL_PORT="5432"
POSTGRESQL_DATABASE="hub"
POSTGRESQL_USER="postgres"
POSTGRESQL_PASSWORD="postgres"
