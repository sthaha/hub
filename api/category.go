package hub

import (
	"context"
	"log"

	category "github.com/tektoncd/hub/api/gen/category"
)

// category service example implementation.
// The example methods log the requests and return zero values.
type categorysrvc struct {
	logger *log.Logger
}

// NewCategory returns the category service implementation.
func NewCategory(logger *log.Logger) category.Service {
	return &categorysrvc{logger}
}

// Get all Categories with their associated tags.
func (s *categorysrvc) Categories(ctx context.Context) (res *category.CategoriesResult, err error) {
	res = &category.CategoriesResult{}
	s.logger.Print("category.Categories")
	return
}
