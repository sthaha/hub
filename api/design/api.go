package design

import (
	. "goa.design/goa/v3/dsl"
	cors "goa.design/plugins/v3/cors/dsl"
)

var _ = API("hub", func() {
	Title("Tekton Hub")
	Description("HTTP services for managing Tekton Hub")
	Version("1.0")
	Meta("swagger:example", "false")
	Server("hub", func() {
		Services("category", "swagger")
		Host("localhost", func() {
			URI("http://localhost:8000")
		})
	})
	cors.Origin("*", func() {
		cors.Headers("Content-Type")
		cors.Methods("GET")
	})
})
