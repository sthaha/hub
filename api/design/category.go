package design

import (
	. "goa.design/goa/v3/dsl"
)

var _ = Service("category", func() {
	Description("The category service gives categories details")

	//Method to get all categories with their tags
	Method("Categories", func() {
		Description("Get all Categories with their associated tags.")
		Result(func() {
			Attribute("data", ArrayOf(Category))
			Attribute("errors", ArrayOf(String)) // To be updated
		})
		HTTP(func() {
			GET("/categories")
			Response(StatusOK)
		})
	})
})