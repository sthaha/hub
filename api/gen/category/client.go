// Code generated by goa v3.1.1, DO NOT EDIT.
//
// category client
//
// Command:
// $ goa gen github.com/tektoncd/hub/api/design

package category

import (
	"context"

	goa "goa.design/goa/v3/pkg"
)

// Client is the "category" service client.
type Client struct {
	CategoriesEndpoint goa.Endpoint
}

// NewClient initializes a "category" service client given the endpoints.
func NewClient(categories goa.Endpoint) *Client {
	return &Client{
		CategoriesEndpoint: categories,
	}
}

// Categories calls the "Categories" endpoint of the "category" service.
func (c *Client) Categories(ctx context.Context) (res *CategoriesResult, err error) {
	var ires interface{}
	ires, err = c.CategoriesEndpoint(ctx, nil)
	if err != nil {
		return
	}
	return ires.(*CategoriesResult), nil
}
