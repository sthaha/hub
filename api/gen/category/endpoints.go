// Code generated by goa v3.1.1, DO NOT EDIT.
//
// category endpoints
//
// Command:
// $ goa gen github.com/tektoncd/hub/api/design

package category

import (
	"context"

	goa "goa.design/goa/v3/pkg"
)

// Endpoints wraps the "category" service endpoints.
type Endpoints struct {
	Categories goa.Endpoint
}

// NewEndpoints wraps the methods of the "category" service with endpoints.
func NewEndpoints(s Service) *Endpoints {
	return &Endpoints{
		Categories: NewCategoriesEndpoint(s),
	}
}

// Use applies the given middleware to all the "category" service endpoints.
func (e *Endpoints) Use(m func(goa.Endpoint) goa.Endpoint) {
	e.Categories = m(e.Categories)
}

// NewCategoriesEndpoint returns an endpoint function that calls the method
// "Categories" of service "category".
func NewCategoriesEndpoint(s Service) goa.Endpoint {
	return func(ctx context.Context, req interface{}) (interface{}, error) {
		return s.Categories(ctx)
	}
}
