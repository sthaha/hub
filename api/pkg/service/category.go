package service

import (
	"context"
	"fmt"

	"github.com/jinzhu/gorm"
	"github.com/tektoncd/hub/api/pkg/db/model"
	"go.uber.org/zap"

	category "github.com/tektoncd/hub/api/gen/category"
	app "github.com/tektoncd/hub/api/pkg/app"
)

// category service example implementation.
// The example methods log the requests and return zero values.
type categorysrvc struct {
	logger *zap.SugaredLogger
	db     *gorm.DB
}

// NewCategory returns the category service implementation.
func NewCategory(api app.Config) category.Service {
	return &categorysrvc{api.Logger(), api.DB()}
}

// List all categories along with their tags sorted by name
func (s *categorysrvc) List(ctx context.Context) (res []*category.Category, err error) {
	var all []model.Category
	if err := s.db.Order("name").Preload("Tags").Find(&all).Error; err != nil {
		s.logger.Error(err)
		return []*category.Category{}, category.MakeInternalError(fmt.Errorf("Failed to fetch categories"))
	}

	for _, c := range all {
		tags := []*category.Tag{}
		for _, t := range c.Tags {
			tags = append(tags, &category.Tag{
				ID:   t.ID,
				Name: t.Name,
			})
		}
		res = append(res, &category.Category{
			ID:   c.ID,
			Name: c.Name,
			Tags: tags,
		})
	}

	return res, nil
}
