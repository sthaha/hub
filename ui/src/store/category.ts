import { types, getEnv, flow, Instance } from 'mobx-state-tree';
import { Api } from '../api';

export const Tag = types.model('Tags', {
  id: types.identifierNumber,
  name: types.string
});

export const Category = types
  .model('Category', {
    id: types.identifierNumber,
    name: types.string,
    tags: types.array(types.reference(Tag)),
    selected: false
  })

  .actions((self) => ({
    toggle() {
      self.selected = !self.selected;
    }
  }));

export type ITag = Instance<typeof Tag>;
export type ICategory = Instance<typeof Category>;
export type ICategoryStore = Instance<typeof CategoryStore>;

export const CategoryStore = types
  .model('CategoryStore', {
    items: types.map(Category),
    tags: types.optional(types.map(Tag), {}),
    isLoading: true,
    err: ''
  })

  .views((self) => ({
    get api(): Api {
      return getEnv(self).api;
    },

    get count() {
      return self.items.size;
    },

    get values() {
      return Array.from(self.items.values());
    },

    get selectedTags() {
      return new Set(
        Array.from(self.items.values())
          .filter((c: ICategory) => c.selected)
          .reduce((acc: number[], c: ICategory) => [...acc, ...c.tags.map((t: ITag) => t.id)], [])
      );
    }
  }))

  .actions((self) => ({
    add(item: ICategory) {
      self.items.put(item);
    },

    setLoading(l: boolean) {
      self.isLoading = l;
    },

    clearSelected() {
      self.items.forEach((c) => {
        c.selected = false;
      });
    }
  }))

  .actions((self) => ({
    load: flow(function* () {
      try {
        self.setLoading(true);
        const { api } = self;

        const json = yield api.categories();

        // adding the tags to the store - normalized
        const tags: ITag[] = json.data.flatMap((item: ICategory) => item.tags);

        tags.forEach((t) => self.tags.put(t));

        // creating the model only after the store has the tags normalized
        const categories: ICategory[] = json.data.map((c: ICategory) => ({
          id: c.id,
          name: c.name,
          tags: c.tags.map((tag: ITag) => tag.id)
        }));

        categories.forEach((c: ICategory) => self.add(c));
      } catch (err) {
        self.err = err.toString();
      }

      self.setLoading(false);
    })
  }))

  .actions((self) => ({
    afterCreate() {
      self.load();
    }
  }));
