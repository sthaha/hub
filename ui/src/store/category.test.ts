import { when } from 'mobx';
import { getSnapshot } from 'mobx-state-tree';
import { FakeHub } from '../api/testutil';
import { CategoryStore, Category, Tag } from './category';

const TESTDATA_DIR = `${__dirname}/testdata`;
const api = new FakeHub(TESTDATA_DIR);

describe('Store Object', () => {
  it('can create a tag object', () => {
    const store = Tag.create({
      id: 1,
      name: 'cli'
    });
    expect(store.name).toBe('cli');
  });
  it('can create a category object', () => {
    const category = Category.create({
      id: 1,
      name: 'test',
      tags: ['1']
    });

    expect(category.name).toBe('test');
    expect(category.tags.length).toBe(1);
  });
});

describe('Store functions', () => {
  it('can create a category store', (done) => {
    const store = CategoryStore.create({}, { api });
    expect(store.count).toBe(0);
    expect(store.isLoading).toBe(true);
    when(
      () => !store.isLoading,
      () => {
        expect(store.count).toBe(5);
        expect(store.isLoading).toBe(false);

        expect(getSnapshot(store)).toMatchSnapshot();

        done();
      }
    );
  });

  it('can toggle the selected category', (done) => {
    const store = CategoryStore.create({}, { api });
    expect(store.count).toBe(0);
    expect(store.isLoading).toBe(true);

    when(
      () => !store.isLoading,
      () => {
        expect(store.count).toBe(5);
        expect(store.isLoading).toBe(false);

        store.categories.get('1')?.toggle();

        expect(store.categories.get('1')?.selected).toBe(true);

        done();
      }
    );
  });

  it('can clear all the categories', (done) => {
    const store = CategoryStore.create({}, { api });
    expect(store.count).toBe(0);

    when(
      () => !store.isLoading,
      () => {
        expect(store.count).toBe(5);
        expect(store.isLoading).toBe(false);

        store.categories.get('1')?.toggle();
        store.categories.get('2')?.toggle();

        store.clearSelected();
        expect(store.categories.get('1')?.selected).toBe(false);

        done();
      }
    );
  });

  it('can return the tags for the categories which are selected', (done) => {
    const store = CategoryStore.create({}, { api });
    expect(store.count).toBe(0);
    expect(store.isLoading).toBe(true);

    when(
      () => !store.isLoading,
      () => {
        expect(store.count).toBe(5);
        expect(store.isLoading).toBe(false);

        store.categories.get('1')?.toggle();
        store.categories.get('2')?.toggle();

        expect(store.selectedTags.size).toBe(2);

        done();
      }
    );
  });
});
