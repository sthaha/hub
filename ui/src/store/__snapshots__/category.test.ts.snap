// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Store functions can create a category store 1`] = `
Object {
  "categories": Object {
    "1": Object {
      "id": 1,
      "name": "Build Tools",
      "selected": false,
      "tags": Array [
        1,
      ],
    },
    "2": Object {
      "id": 2,
      "name": "CLI",
      "selected": false,
      "tags": Array [
        2,
      ],
    },
    "3": Object {
      "id": 3,
      "name": "Cloud",
      "selected": false,
      "tags": Array [
        4,
        5,
        6,
        3,
      ],
    },
    "4": Object {
      "id": 4,
      "name": "Deploy",
      "selected": false,
      "tags": Array [
        7,
      ],
    },
    "5": Object {
      "id": 5,
      "name": "Image Build",
      "selected": false,
      "tags": Array [
        8,
      ],
    },
  },
  "err": "",
  "isLoading": false,
  "tags": Object {
    "1": Object {
      "id": 1,
      "name": "build-tool",
    },
    "2": Object {
      "id": 2,
      "name": "cli",
    },
    "3": Object {
      "id": 3,
      "name": "gcp",
    },
    "4": Object {
      "id": 4,
      "name": "aws",
    },
    "5": Object {
      "id": 5,
      "name": "azure",
    },
    "6": Object {
      "id": 6,
      "name": "cloud",
    },
    "7": Object {
      "id": 7,
      "name": "deploy",
    },
    "8": Object {
      "id": 8,
      "name": "image-build",
    },
  },
}
`;
